File: src/client/Smartcomply.ts
import { SDKConfig } from "./Config";
import { HttpClient } from "./HttpClient";
import { OnboardingModule } from "../modules/onboarding/onboarding";
import { LivenessModule } from "../modules/liveness/liveness";
import {
  SessionResponse,
  SDKInitConfig,
} from "../types/onboarding";

/**
 * SmartComply SDK — main entry point.
 *
 * Usage:
 *   const sdk = new SmartComply({ apiKey: "pk_...", clientId: "uuid-...", environment: "sandbox" });
 *   const session = await sdk.createSession();
 *   const config = await sdk.initializeConfig();
 *   // ... use sdk.onboarding and sdk.liveness modules
 */
export class SmartComply {
  private http: HttpClient;
  private clientId: string;
  private _sessionToken: string | null = null;
  private _sessionExpiresAt: string | null = null;
  private _sdkConfig: SDKInitConfig | null = null;

  public onboarding: OnboardingModule;
  public liveness: LivenessModule;

  constructor(config: SDKConfig) {
    if (!config.apiKey) {
      throw new Error("SmartComply: apiKey is required");
    }
    if (!config.clientId) {
      throw new Error("SmartComply: clientId is required (UUID from your SDK config)");
    }

    this.clientId = config.clientId;
    this.http = new HttpClient(config);

    this.onboarding = new OnboardingModule(this.http);
    this.liveness = new LivenessModule(this.http);
  }

  /** Current session token (null if not created yet) */
  get sessionToken(): string | null {
    return this._sessionToken;
  }

  /** Session expiration ISO timestamp */
  get sessionExpiresAt(): string | null {
    return this._sessionExpiresAt;
  }

  /** SDK config fetched from backend (null until initializeConfig() is called) */
  get sdkConfig(): SDKInitConfig | null {
    return this._sdkConfig;
  }

  /**
   * Create a new SDK session.
   *
   * POST /v1/sdk/session/create/
   * Auth: Authorization: Bearer <apiKey>
   * Body: { client_id: "<uuid>" }
   *
   * Returns: { token: "...", expires_at: "..." }
   */
  async createSession(): Promise<SessionResponse> {
    const response = await this.http.request<SessionResponse>(
      "POST",
      "/v1/sdk/session/create/",
      { client_id: this.clientId }
    );

    this._sessionToken = response.token;
    this._sessionExpiresAt = response.expires_at;

    // Set the session token on the HTTP client for subsequent requests
    this.http.setSessionToken(response.token);

    return response;
  }

  /**
   * Fetch SDK configuration from the backend.
   * Must be called after createSession().
   *
   * GET /v1/sdk/sdk/initialize/
   * Auth: x-access-token: <sessionToken>
   *
   * Returns: { brand_name, theme, verification_type, channels, redirect_url, ... }
   */
  async initializeConfig(): Promise<SDKInitConfig> {
    const config = await this.http.sessionRequest<SDKInitConfig>(
      "GET",
      "/v1/sdk/sdk/initialize/"
    );

    this._sdkConfig = config;
    return config;
  }
}

File: flutter/lib/src/client/smartcomply.dart
import 'http_client.dart';
import 'config.dart';
import '../modules/onboarding/onboarding_module.dart';
import '../modules/liveness/liveness_module.dart';
import '../types/onboarding.dart';
import '../types/liveness.dart';

/// The primary entry point for the SmartComply Flutter SDK.
///
/// Usage:
/// ```dart
/// final sdk = SmartComply(SDKConfig(
///   apiKey: 'your_api_key',
///   clientId: 'your-client-uuid',
///   environment: Environment.production,
/// ));
///
/// await sdk.createSession();
/// final identity = await sdk.onboarding.verify(...);
/// final liveness = await sdk.liveness.startCheck(context, ...);
/// ```
class SmartComply {
  final HttpClient _http;
  final String _clientId;
  String? _sessionToken;

  late final OnboardingModule onboarding;
  late final LivenessModule liveness;

  SmartComply(SDKConfig config)
      : _http = HttpClient(config),
        _clientId = config.clientId {
    if (config.apiKey.isEmpty) {
      throw ArgumentError('SmartComply: apiKey is required');
    }
    if (config.clientId.isEmpty) {
      throw ArgumentError('SmartComply: clientId is required');
    }
    onboarding = OnboardingModule(_http);
    liveness = LivenessModule(_http);
  }

  /// The active session token, or null if [createSession] has not been called.
  String? get sessionToken => _sessionToken;

  /// Creates a new verification session.
  ///
  /// Uses the [apiKey] for this one call. After success, all subsequent
  /// SDK requests automatically use the returned session token.
  Future<SessionResponse> createSession() async {
    // Session creation must use the API key, not a session token
    final response = await _http.request<SessionResponse>(
      'POST',
      '/v1/session/create',
      body: {'client_id': _clientId},
      useApiKey: true,
      fromJson: SessionResponse.fromJson,
    );

    _sessionToken = response.token;

    // Switch the HttpClient to use the session token for all future calls
    _http.setSessionToken(response.token);
    onboarding.setSessionId(response.token);
    liveness.setSessionId(response.token);

    return response;
  }

  /// Convenience: creates a session (if none exists) + verifies identity.
  Future<VerifyIdentityResponse> startOnboarding({
    required OnboardingType onboardingType,
    required String idNumber,
  }) async {
    if (_sessionToken == null) await createSession();
    return onboarding.verify(
      onboardingType: onboardingType,
      idNumber: idNumber,
    );
  }

  /// Fetches the SDK configuration (brand name, theme, channels) from the
  /// backend after a session is created.
  Future<SDKInitializeResponse> initialize() async {
    if (_sessionToken == null) {
      throw StateError('Call createSession() before initialize().');
    }
    return _http.request<SDKInitializeResponse>(
      'GET',
      '/v1/sdk/initialize',
      fromJson: SDKInitializeResponse.fromJson,
    );
  }
}
